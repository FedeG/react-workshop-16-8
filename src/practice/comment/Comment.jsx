import React from 'react';
import PropTypes from 'prop-types';
import Loadable from 'react-loadable';

import Avatar from './Avatar';
import UserInfo from './UserInfo';
import CommentText from './CommentText';
import CommentDate from './CommentDate';

const LoadableClock = Loadable({
  loader: () => import('../clock'),
  loading: () => <div>Loading...</div>,
});

export default class Comment extends React.PureComponent {
  static propTypes = {
    author: PropTypes.shape({
      avatarUrl: PropTypes.string,
      name: PropTypes.string,
    }).isRequired,
    date: PropTypes.instanceOf(Date).isRequired,
    comments: PropTypes.arrayOf(PropTypes.string),
  };

  static defaultProps = {
    comments: [],
  };

  state = {
    date: new Date(),
    showDate: false,
  }

  onClickHandler = () => {
    this.setState({ date: new Date(), showDate: true });
  }

  getComments() {
    const { comments } = this.props;
    return comments.map(comment => <h4 key={comment}>{comment}</h4>);
  }

  render() {
    const { date: clockDate, showDate } = this.state;
    const { author, date } = this.props;
    return (
      <div className="Comment">
        <UserInfo author={author} onClickHandler={this.onClickHandler} />
        <CommentText>
          <div>
            <h1>Comentarios:</h1>
            <Avatar author={author} onClickHandler={this.onClickHandler} />
            {this.getComments()}
          </div>
        </CommentText>
        <CommentDate date={date} />
        { showDate && <LoadableClock date={clockDate} />}
      </div>
    );
  }
}
