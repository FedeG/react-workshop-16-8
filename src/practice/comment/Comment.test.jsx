import React from 'react';
import MockDate from 'mockdate';
import renderer from 'react-test-renderer';

import Comment from './Comment';

jest.mock('../clock', () => 'Clock');
jest.mock('./Avatar', () => 'Avatar');
jest.mock('./UserInfo', () => 'UserInfo');
jest.mock('./CommentText', () => 'CommentText');
jest.mock('./CommentDate', () => 'CommentDate');

describe('Comment', () => {
  let component;

  const author = {
    name: 'Test',
    avatarUrl: 'avatarTestUrl',
  };

  beforeEach(() => {
    MockDate.set('3/5/1995');
    component = renderer.create(
      <Comment
        author={author}
        date={new Date()}
        comments={['buenos dias', 'buenas tarde', 'buenas noches']}
      />,
    );
  });

  test('initial state', () => {
    const instance = component.getInstance();
    expect(instance.state.date).toEqual(new Date());
    expect(instance.state.showDate).toBeFalsy();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('when click, showDate should be true', () => {
    const instance = component.getInstance();
    expect(instance.state.showDate).toBeFalsy();

    instance.onClickHandler();

    expect(instance.state.showDate).toBeTruthy();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('when click, date should change', () => {
    const instance = component.getInstance();
    expect(instance.state.date).toEqual(new Date());

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    MockDate.set('3/5/1996');

    instance.onClickHandler();

    expect(instance.state.date).toEqual(new Date());

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    MockDate.set('3/5/1997');

    instance.onClickHandler();

    expect(instance.state.date).toEqual(new Date());

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
