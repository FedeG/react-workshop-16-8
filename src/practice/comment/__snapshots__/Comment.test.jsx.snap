// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment initial state 1`] = `
<div
  className="Comment"
>
  <UserInfo
    author={
      Object {
        "avatarUrl": "avatarTestUrl",
        "name": "Test",
      }
    }
    onClickHandler={[Function]}
  />
  <CommentText>
    <div>
      <h1>
        Comentarios:
      </h1>
      <Avatar
        author={
          Object {
            "avatarUrl": "avatarTestUrl",
            "name": "Test",
          }
        }
        onClickHandler={[Function]}
      />
      <h4>
        buenos dias
      </h4>
      <h4>
        buenas tarde
      </h4>
      <h4>
        buenas noches
      </h4>
    </div>
  </CommentText>
  <CommentDate
    date={1995-03-05T03:00:00.000Z}
  />
</div>
`;

exports[`Comment when click, date should change 1`] = `
<div
  className="Comment"
>
  <UserInfo
    author={
      Object {
        "avatarUrl": "avatarTestUrl",
        "name": "Test",
      }
    }
    onClickHandler={[Function]}
  />
  <CommentText>
    <div>
      <h1>
        Comentarios:
      </h1>
      <Avatar
        author={
          Object {
            "avatarUrl": "avatarTestUrl",
            "name": "Test",
          }
        }
        onClickHandler={[Function]}
      />
      <h4>
        buenos dias
      </h4>
      <h4>
        buenas tarde
      </h4>
      <h4>
        buenas noches
      </h4>
    </div>
  </CommentText>
  <CommentDate
    date={1995-03-05T03:00:00.000Z}
  />
</div>
`;

exports[`Comment when click, date should change 2`] = `
<div
  className="Comment"
>
  <UserInfo
    author={
      Object {
        "avatarUrl": "avatarTestUrl",
        "name": "Test",
      }
    }
    onClickHandler={[Function]}
  />
  <CommentText>
    <div>
      <h1>
        Comentarios:
      </h1>
      <Avatar
        author={
          Object {
            "avatarUrl": "avatarTestUrl",
            "name": "Test",
          }
        }
        onClickHandler={[Function]}
      />
      <h4>
        buenos dias
      </h4>
      <h4>
        buenas tarde
      </h4>
      <h4>
        buenas noches
      </h4>
    </div>
  </CommentText>
  <CommentDate
    date={1995-03-05T03:00:00.000Z}
  />
  <Clock
    date={1996-03-05T03:00:00.000Z}
  />
</div>
`;

exports[`Comment when click, date should change 3`] = `
<div
  className="Comment"
>
  <UserInfo
    author={
      Object {
        "avatarUrl": "avatarTestUrl",
        "name": "Test",
      }
    }
    onClickHandler={[Function]}
  />
  <CommentText>
    <div>
      <h1>
        Comentarios:
      </h1>
      <Avatar
        author={
          Object {
            "avatarUrl": "avatarTestUrl",
            "name": "Test",
          }
        }
        onClickHandler={[Function]}
      />
      <h4>
        buenos dias
      </h4>
      <h4>
        buenas tarde
      </h4>
      <h4>
        buenas noches
      </h4>
    </div>
  </CommentText>
  <CommentDate
    date={1995-03-05T03:00:00.000Z}
  />
  <Clock
    date={1997-03-05T03:00:00.000Z}
  />
</div>
`;

exports[`Comment when click, showDate should be true 1`] = `
<div
  className="Comment"
>
  <UserInfo
    author={
      Object {
        "avatarUrl": "avatarTestUrl",
        "name": "Test",
      }
    }
    onClickHandler={[Function]}
  />
  <CommentText>
    <div>
      <h1>
        Comentarios:
      </h1>
      <Avatar
        author={
          Object {
            "avatarUrl": "avatarTestUrl",
            "name": "Test",
          }
        }
        onClickHandler={[Function]}
      />
      <h4>
        buenos dias
      </h4>
      <h4>
        buenas tarde
      </h4>
      <h4>
        buenas noches
      </h4>
    </div>
  </CommentText>
  <CommentDate
    date={1995-03-05T03:00:00.000Z}
  />
  <div>
    Loading...
  </div>
</div>
`;
